### PR TITLE
Tabulated fluid property interface for (v,e) & (v,h) lookups

### PR DESCRIPTION
**Pull request by snschune**
refs #20101 

## Reason

Tabulated properties for MOOSE TH applications that use `(v,e), (v,h)` formulations.

## Design

Create reverse interpolation for `p T of (v,e) or (v,h)`.

## Impact

no

### CODE REVIEW

File: modules/fluid_properties/unit/src/TabulatedFluidPropertiesTest.C [hunk 1]
```diff
@@ -303,6 +303,32 @@ TEST_F(TabulatedFluidPropertiesTest, fromFileVE)
...
+    Moose::derivInsert(dvdx, 1, 0);
+    Moose::derivInsert(dedx, 0, 0);
+    Moose::derivInsert(dedx, 1, 1);
+
+    DualReal v_ad(v, dvdx);
+    DualReal e_ad(e, dedx);
+    // this line does not work
+    // error: no matching member function for call to 'p_from_v_e'
+    // candidate function not viable: no known conversion from 'DualReal' (aka 'DualNumber<double, SemiDynamicSparseNumberArray<double, unsigned long long, NWrapper<50>>, true>') to 'libMesh::Real' (aka 'double') for 1st argument
+    //DualReal p_ad = _tab_fp_ve->p_from_v_e(v_ad, e_ad);
```
Comment (snschune) [inline modules/fluid_properties/unit/src/TabulatedFluidPropertiesTest.C:325]: This line @andrsd

File: modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h [hunk 2]
```diff
@@ -236,6 +277,49 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
 
   /// The MOOSE delimited file reader.
   MooseUtils::DelimitedFileReader _csv_reader;
+
+  /// if the loopup table p(v, e) and T(v, e) should be constructed
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:281]: "loopup" -> "lookup"

File: modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h [hunk 3]
```diff
@@ -142,6 +153,34 @@ class TabulatedFluidProperties : public SinglePhaseFluidProperties
 
   virtual void vaporPressure(Real temperature, Real & psat, Real & dpsat_dT) const override;
 
+  /**
+   * Derivatives like dc_dv & dc_de are computed using the chain role
```
Comment (andrsd) [inline modules/fluid_properties/include/userobjects/TabulatedFluidProperties.h:157]: "role" -> "rule"

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C [hunk 4]
```diff
@@ -51,6 +52,29 @@ TabulatedFluidProperties::validParams()
                                   properties,
                                   "Properties to interpolate if no data file is provided");
   params.addParam<bool>("save_file", true, "Whether to save the csv fluid properties file");
+  params.addParam<bool>("construct_pT_from_ve",
+                        false,
+                        "If the loopup table (p, T) as functions of (v, e) should be constructed.");
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:57]: "loopup" -> "lookup"

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C [hunk 5]
```diff
@@ -313,6 +497,38 @@ TabulatedFluidProperties::molarMass() const
...
+  if (_interpolate_density)
+  {
+    checkInputVariables(pressure, temperature);
+    _property_ipol[_density_idx]->sampleValueAndDerivatives(
+        pressure, temperature, rho, drho_dp, drho_dT);
+  }
+  else
+    _fp.rho_from_p_T(pressure, temperature, rho, drho_dp, drho_dT);
+
+  // convert from rho to v
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:526]: cpt. Obvious -> delete the comment.

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C [hunk 6]
```diff
@@ -491,6 +740,43 @@ TabulatedFluidProperties::s_from_p_T(Real p, Real T, Real & s, Real & ds_dp, Rea
...
 }
 
+Real
+TabulatedFluidProperties::g_from_v_e(Real v, Real e) const
+{
+  const Real p = p_from_v_e(v, e);
+  const Real T = T_from_v_e(v, e);
+  const Real s = s_from_p_T(p, T);
+  const Real h = h_from_p_T(p, T);
+  // this is simple the definition of Gibbs free energy
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:750]: simple -> simply

File: modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C [hunk 7]
```diff
@@ -491,6 +740,43 @@ TabulatedFluidProperties::s_from_p_T(Real p, Real T, Real & s, Real & ds_dp, Rea
...
+  return h - T * s;
+}
+
+Real
+TabulatedFluidProperties::e_from_v_h(Real v, Real h) const
+{
+  if (!_construct_pT_from_ve)
+    mooseError("You must construct pT from ve tables when calling e_from_v_h.");
```
Comment (andrsd) [inline modules/fluid_properties/src/userobjects/TabulatedFluidProperties.C:758]: maybe use `(p, T)` and `(v, e)` instead of `pT` and `ve`. Or `pt-` and `ve-tables`. It may not be obvious to a user what you are trying to tell them. Also mention, that you want them to use the `construct_pT_from_ve` param set to `true`. Otherwise this error message is fairly useless.

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by snschune** [comment]
@andrsd I noted that I had problems with this FP object when I used it in THM. 
I investigated a bit more and among the issues I found is that `v(p, T)` and `e(p, T)`
when interpolated by the splines is not monotonic even if the original tabulation points are.

This is an issue because I invert `(p, T) -> (e, v)` and if that is not a bi-monotonic map, then
I cannot uniquely invert it. Of course thermodynamics tells that both `(p, T)` and `(v,e)` specify
unique states so the non-monotonicity of the interpolation violates basic physics. 

In summary, I don't know if this will ever work the way it's implemented right now. We could go
to guaranteed monotonic interpolations like linear interpolation, but at this point, I am not sure
if it's worth pursuing.

**Comment by snschune** [comment]
This will never worked as discussed with @andrsd
